numba/core/interpreter.py

```diff
@@ -79,6 +79,647 @@ def get_assignment_source(self, destname):
+                        #  $arg1 = ...
+                        #  $arg1_tup = build_tuple(items=[$arg1])
+                        #  $combo_tup_2 = $combo_tup_1 + $arg1_tup
+                        #  ...
+                        #  $combo_tup_n = $combo_tup_{n-1} + $argn_tup
+                        #
+                        # In addition, the IR seems to contain a final
+                        # assignment for the varargs that looks like:
+                        #
+                        #  $varargs_var = $combo_tup_n
+                        #
+                        # Here args_def is expected to be a simple assignment.
+                        # However, it is unclear if this extra assignment is
+                        # always generated, so as a result the code is written
+                        # to support args_def being either $varargs_var
+                        # or $combo_tup_n from the above example.
+                        args = _call_function_ex_replace_args_large(
+                            blk.body,
+                            args_def,
+                            new_body,
+                            vararg_loc,
+                            func_ir,
+                            errmsg,
+                        )
+                # Create a new call updating the args and kws
+                new_call = ir.Expr.call(
+                    call.func, args, kws, call.loc, target=call.target
+                )
+                # Drop the existing definition for this stmt.
+                _remove_assignment_definition(blk.body, i, func_ir)
+                # Update the statement
+                stmt = ir.Assign(new_call, stmt.target, stmt.loc)
+                # Update the definition
+                func_ir._definitions[stmt.target.name].append(new_call)
+            elif (
+                isinstance(stmt, ir.Assign)
+                and isinstance(stmt.value, ir.Expr)
+                and stmt.value.op == "call"
+                and stmt.value.vararg is not None
+            ):
+                # If there is a call with vararg we need to check
+                # if the list -> tuple conversion failed and if so
+                # throw an error.
+                call = stmt.value
+                args = call.args
+                vararg = call.vararg
+                if args:
+                    # If we have vararg then args is expected to
+                    # be an empty list.
+                    raise UnsupportedError(errmsg)
+                vararg_loc = i - 1
+                args_def = None
+                not_found = True
+                while vararg_loc >= 0 and not_found:
+                    args_def = blk.body[vararg_loc]
+                    if (
+                        isinstance(args_def, ir.Assign)
+                        and args_def.target.name == vararg.name
+                    ):
+                        not_found = False
+                    else:
+                        vararg_loc -= 1
```

Comment:
(Leaving thought trail in case we need it in the future. This probably doesn't present to be a problem)

With bytecode from:

```python
args = (*(0, 1, 2),)   # the starunpack triggers LIST_TO_TUPLE
y = my_func2(*args)
```
```
 16           0 BUILD_LIST               0
              2 LOAD_CONST               1 ((0, 1, 2))
              4 LIST_EXTEND              1
              6 LIST_TO_TUPLE
              8 STORE_FAST              31 (args)

 17          10 LOAD_GLOBAL              0 (my_func2)
             12 LOAD_FAST               31 (args)
             14 CALL_FUNCTION_EX         0
             16 STORE_FAST              32 (y)
```
the code here scan back from the call to `my_func2` to find the definition of `args`, which is suppose to be a `LIST_TO_TUPLE` opcode, but is replaced by `peep_hole_list_to_tuple`; otherwise, the later check will reject this call.

Reply:
I have also tried to trick this part of the code with the walrus operator but Numba is already miscompiling this on mainline:

```python
my_func2(args:=(*(0, 1, 2),), *args)
```